rust/driver-ipc/src/client.rs

```diff
@@ -1,180 +1,136 @@
-        Self::connect_to("virtualdisplaydriver").await
+    /// Connect to driver on pipe with default name.
+    ///
+    /// The default name is [DEFAULT_PIPE_NAME].
+    pub fn connect() -> Result<Self> {
+        Self::connect_to(DEFAULT_PIPE_NAME)
     }
 
-    // choose which pipe name you connect to
-    // pipe name is ONLY the name, only the {name} portion of \\.\pipe\{name}
-    pub async fn connect_to(name: &str) -> Result<Self> {
-        let client = ClientOptions::new()
+    /// Connect to driver on pipe with specified name.
+    ///
+    /// `name` is ONLY the {name} portion of \\.\pipe\{name}.
+    pub fn connect_to(name: &str) -> Result<Self> {
+        let client = named_pipe::ClientOptions::new()
             .read(true)
             .write(true)
-            .pipe_mode(PipeMode::Byte)
+            .pipe_mode(named_pipe::PipeMode::Byte)
             .open(format!(r"\\.\pipe\{name}"))
             .map_err(IpcError::ConnectionFailed)?;
+        let client = Arc::new(client);
 
-        let client = client;
-        let user_is_event = AtomicBool::new(false);
-        let (event_send, event_recv) = unbounded_channel();
-        let (client_send, client_recv) = unbounded_channel();
-
-        let inner = Inner {
-            client,
-            user_is_event,
-            event_recv: Mutex::new(event_recv),
-            client_recv: Mutex::new(client_recv),
-        };
-
-        let client = Self(Arc::new(inner));
-
-        let (tx, mut rx) = unbounded_channel();
+        let notify = Arc::new(Notify::new());
 
-        // receive command thread
-        let client2 = client.clone();
-        task::spawn(async move {
-            _ = receive_command(&client2, tx).await;
-        });
-
-        let client2 = client.clone();
-        task::spawn(async move {
-            let user_is_event = &client2.0.user_is_event;
-
-            loop {
-                let Some(data) = rx.recv().await else {
-                    // sender closed
-                    break;
-                };
-
-                let user_is_event = user_is_event.load(Ordering::Acquire);
-                let data_is_event = matches!(data, ClientCommand::Event(_));
-
-                if user_is_event && data_is_event {
-                    let ClientCommand::Event(e) = data else {
-                        unreachable!()
-                    };
+        let (command_tx, _command_rx) = broadcast::channel::<ClientCommand>(10);
 
-                    if user_is_event {
-                        _ = event_send.send(e);
-                    }
-                } else if let ClientCommand::Reply(r) = data {
-                    _ = client_send.send(r);
-                }
-            }
-        });
+        {
+            let client = client.clone();
+            let command_tx = command_tx.clone();
+            let notify = notify.clone();
+            task::spawn(async move { receive_command(&client, command_tx, &notify).await });
```

Comment:
I know I didn't have this in the original code, but I think we should handle the case where this errors out. Basically, what I'm envisioning here is that read errors out, but it's then confusing to the library user on why they aren't receiving any messages (e.g. on their event stream).

The senders should be fine since they'll keep receiving errors if they try to send, but since we still hold onto a broadcast subscriber, the subscribers won't receive a `None`.

I'm thinking that  we should completely drop the broadcast sender on error to solve this.

Reply:
Whether the callbacks hung or not would be invisible to the user anyways. Still I'm wondering if maybe we should add a flag, callback, or something somewhere to indicate the connection status of the client.

Reply:
These are error types I would propose.

```rust
pub mod error {
    use thiserror::Error;
    use super::*;

    /// Error returned from [Client::connect] and [Client::connect_to].
    #[derive(Debug, Error)]
    pub enum ConnectionError {
        #[error("Failed to open pipe: {0}")]
        Failed(io::Error),
    }

    /// Error returned from [send_command]
    #[derive(Debug, Error)]
    pub(super) enum SendCommandError {
        #[error("Failed to encode message: {0}")]
        Encode(serde_json::Error),
        #[error("Failed to send message: {0}")]
        PipeBroken(io::Error),
    }

    /// Error returned from [Client::notify], [Client::remove] and
    /// [Client::remove_all].
    #[derive(Debug, Error)]
    pub enum SendError {
        #[error("Failed to send message: {0}")]
        PipeBroken(io::Error),
    }

    /// Error returned from [Client::request_state].
    #[derive(Debug, Error)]
    pub enum RequestError {
        #[error("Failed to send message (pipe broken): {0}")]
        Send(io::Error),
        #[error("Failed to receive message (pipe broken): {0}")]
        Receive(io::Error),
        #[error("Did not get a response in time ({0:?})")]
        Timeout(Duration),
    }

    // TODO: Client::receive_events()

    /// Error returned from [Client::persist].
    #[derive(Debug, Error)]
    pub enum PersistError {
        #[error("Failed to open registry key: {0}")]
        Open(io::Error),
        #[error("Failed to set registry value: {0}")]
        Set(io::Error),
        #[error("Failed to serialize monitors: {0}")]
        Serialize(serde_json::Error),
    }
}
```

Reply:
One problem: We cannot easily distribute errors from `send_command()`, since `io::Error` is not copy.

Reply:
> One option is to wrap the event that is send to receivers into a Result. So command_tx would have type Sender<Result<ClientCommand, SomeErrorType>>, receive_events() would return impl Stream<Item = Result<EventCommand, SomeErrorType>>.

I agree with this. It would allow us to send an error to all receivers once the pipe breaks, and we could easily resume `receive_command` again if the client asks to reconnect. The implementation should be real easy to do, and it's clear what has happened since the error was propagated to clients.

> One problem: We cannot easily distribute errors from send_command(), since io::Error is not Clone.

Looks like we can work around this by wrapping it in a `Rc` or `Arc` at least

> But on that note, we should maybe rethink the error system to return specialized errors instead of the general IpcError, since it does not convey what can actually happen. For example, request_state() will never return QueryNotFound error, but the user must still provide a valid path for it.

I also think we should separate these into more specific errors according to the error conditions that can happen (much less superfluous when handling error types for the user).